enarx-keep-sgx-shim/src/start.S

```diff
@@ -5,10 +5,67 @@
 #define STCK    (8 * 1024 * 1024)
 #define HEAP    (128 * 1024 * 1024)
 
-    .section ".enarx.ctx", "aw"
-    .globl ctx
-    .type ctx, @object
-ctx:
+#define GPR     (4096 - 184)
+#define RSP     (GPR + 32)
+
+#define MISC    (GPR - 16)
+#define SRSP    (MISC - 8)
+
+# Clear all preserved (callee-saved) registers (except %rsp)
+.macro  zerop
+    xor     %rbx,                   %rbx
+    xor     %rbp,                   %rbp
+    xor     %r12,                   %r12
+    xor     %r13,                   %r13
+    xor     %r14,                   %r14
+    xor     %r15,                   %r15
+.endm
+
+# Clear all argument registers
+.macro  zeroa
+    xor     %rcx,                   %rcx
+    xor     %rdx,                   %rdx
+    xor     %rsi,                   %rsi
+    xor     %rdi,                   %rdi
+    xor     %r8,                    %r8
+    xor     %r9,                    %r9
+.endm
+
+# Clear all temporary registers
+.macro  zerot
+    xor     %r10,                   %r10
+    xor     %r11,                   %r11
+.endm
+
+# Clear CPU flags using the supplied register (which MUST contain zero!)
+.macro  zerof reg
```

Comment:
Note that this macro currently does _not_ clear all EFLAGS bits. For instance one bit that remains under attacker control afais is the EFLAGS.AC bit which can be abused for side-channel purposes. As it should be fairly trivial to clear this bit, I strongly feel it is the resp of the trusted runtime to do so.

Also I'd advise to check out our Taleof2Worlds paper to try and sanitize the ABI systematically:

https://people.cs.kuleuven.be/~jo.vanbulck/ccs19-tale.pdf